### DISCUSSION TITLE
「パターンの編集ができる」スレ

### DISCUSSION THREAD

**Discussion by msntts** [msntts/PatternLanguageEditor#10] [General]
プロジェクトは[ここ](https://github.com/users/msntts/projects/3/views/12)  
ブランチは[これ](https://github.com/msntts/PatternLanguageEditor/tree/edit-pattern)

**Comment by msntts** [comment]
パッケージを追加したので、 起動の前に ```npm install``` がいるかもしれないです。
とりあえず、編集できそうな画面と挙動を作ってみました👀
ContainerとComponentの考え方についてちょっと考えさせられる感じになってきました。
Componentって今いるかな？っていう感じ。。
その他リファクタリングアイディアとうとうあれば

**Reply by msntts** [reply]
今回の学びとしてはRouter/Link(画面の切り替えに使った)とローカルステート(useState())です。
その辺の不明点とかもあれば聞いてください。

**Reply by kinakomochigit** [reply]
動作を確認しました。使い方の学習を踏まえUseStateをまとめて実装してみました。確認お願いします。

**Reply by msntts** [reply]
UseState()をまとめた版の追加修正しました。[差分](https://github.com/msntts/PatternLanguageEditor/commit/85d052a9651a05cdea098acd1e9dba530bbc4ba7)
変更量多く見えますが、同じような型定義をしている箇所をまとめたのが主な変更です。
(typeは&で定義済みの型に追加することができるみたい💪)

pushしてくれたコードからスプレッド演算子(...って書くやつ)の意味を理解できたので、ほかのところもコード圧縮しました。

あとは、interfaceとtype、functionとアロー関数(()=>{}で定義する奴)みたいな似て非なるものが混在していたのを統一しました。
今のところ、このアプリでは、特段理由がない場合はtype、アロー関数を使用する方針にしました👀

後はnoteにもかいた、connect()非推奨問題対応としてconnect()使わない版に実装を変更しました。
差分([パターンビュー](https://github.com/msntts/PatternLanguageEditor/commit/a78c4562b51b10a42e1c54312ed15bda2b2e45ac)、[パターンリスト](https://github.com/msntts/PatternLanguageEditor/commit/47e88010f6cb2996c7bf284645905374744f6d1f)、[パターンエディット](https://github.com/msntts/PatternLanguageEditor/commit/e8f5976679bbccd3b12dcf9ea32b10dee949ff15))

これを機にContainerとComponentの実装も見直しています。Componentはそもそもで、共通的になUIをくくりだしてPropsを切り替えるだけ～としたかったのですが、今のところEdit以外は一回しか使わなさそうなのでContainerに統合しました。(ContainerとComponentで同じようなコードを書いていたのでかなりすっきりしました🫥)
Editは新規パターン作成の時に使いまわす想定

**Reply by kinakomochigit** [reply]
やっと、、だいたいの流れは理解しました。PatternEditComponent.tsxのPatternEditではローカルステートを使っていますが、どうしてローカルステートを使うのかまでは理解できていません。ステートの管理やとreact要素の階層図的なのをまとめてみようかと思っています。🧐

**Reply by msntts** [reply]
`どうしてローカルステートを使うのかまでは理解できていません`
この画面では、テキストボックスに入力された値を編集完了時にモデルに反映する必要があります。  
入力された文字列をinputタグからgetElementBy~()などで取得し、actionに渡してもいいとはおもうのですがタグを変えた際にactionの呼び出し箇所を変えたりするのが嫌だなーと思ったので、inputタグと変数をデータバインディングしてactionからは何からデータを入力されたのかを意識しないようにしています。  
そこで、変数としてはstoreかローカルstateのどっちを使うの？っていう話があるんですが、storeはアプリケーション全体から参照可能なデータを入れるものだと考えています。(逆に言うと、ほかの画面などから見てほしいデータを入れたい)  
そこで、編集画面を思うと、入力情報ってキャンセルされるかもしれないし、途中で戻るボタン押されるかもしれないし、、、とかでほかの画面から見れてうれしい情報はあまりないように思います🤔
一方、ローカルstateはその画面の中だけで利用されるデータなのでこちらを使っています。
P.S. この回答を書くためにちょっと調べたんですが、、、今回の場合はRefsを使うのが最も限りなく正解に近い気がしました。。
https://react.dev/learn/manipulating-the-dom-with-refs
https://react.dev/learn/referencing-values-with-refs
このあたりですね。
